Solutions/VMware vCenter/Parsers/vCenter.txt

```diff
@@ -22,63 +22,21 @@
-eventtype = "UserLoginSessionEvent",
-eventid = "592838"
-| project-away logindetails,raw_substring;
-let RoleDetails = vcenter_CL
-| where Message has "Role"
-| extend Parser = extract_all(@"^(\S+\s+\S+\s+\S+\s+\S+)\s+(\d+-\d+-\d+T*\s*\d+:\d+:\d+(.\d{3}|.\d+\+\d+:\d+))\s+(\S+)\s(vpxd|vcenter-server:)(\s+(\d+)\s- -\sEvent\s+\[(\d+)\]\s+\[1-1\]\s+\[(\d+-\d+-\d+T\d+:\d+:\d+.\d+Z)\]\s+\[(\S+)\]\s+\[(\w+)\]\s+\[(\S*)]\s+\[(\S*)]\s+\[(\d+)])*\s+\[*((\S*\s*)*)\]*", dynamic([1,2,4,6,7,8,9,10,11,12,13,14,15]),Message)
-| extend eventTime = todatetime(Parser[0][1]),
-hostname = tostring(Parser[0][2]),
-raw_substring = tostring(Parser[0][12])
-| project-away Parser;
-let RoleCreated = RoleDetails
-| where raw_substring has ("created")
-| extend RoleCreated = tostring(split(tostring(raw_substring)," ",2)[0]),
-eventtype = "RoleCreated",
-eventid = "580333"
-| project-away raw_substring;
-let RoleRemoved = RoleDetails
-| where Message has ("removed") and Message !has "modified"
-| extend RoleRemoved = tostring(split(tostring(raw_substring)," ",2)[0]), //Extract the role. currently it is an array of strings
-eventtype = "RoleRemoved",
-eventid = "580349"
-| project-away raw_substring;
-let RoleModified =
-RoleDetails
-| where raw_substring has ("modified")
-| extend RoleModified = extract_all(@"^(Role\s+modified\s*\S*Previous name:\s+(\w+),\s+new\s+name\s+(\w+)\s*\S*Added\s+privileges:\s+(((\w+.\w+\,*)\s*)*)\s*\S*Removed\s+privileges:\s+(((\w+.\w+\,*)\s*)*))", dynamic([2,3,4,7]),raw_substring)
-| extend eventtype = "RoleModified",
-oldrolename = tostring(RoleModified[0][0]),
-newrolename = tostring(RoleModified[0][1]),
-addedPrivilege = split(tostring(RoleModified[0][2]),", "),
-removedPrivilege = split(tostring(RoleModified[0][3]),", "),
-eventid = "580348"
-| project-away raw_substring;
-union AdminLogoutdetails,AdminLoginDetails,RoleRemoved,RoleCreated,RoleModified
\ No newline at end of file
+let vCenter_Login =() {
+vcenter_CL
+| where Message has ("UserLoginSessionEvent")
+| parse Message  with * "Event [" EventId:string "] [1-1] [" EventTime:datetime "] [" EventType:string "] [" EventSeverity "]" * "[User " DomainName:string "\\" Username:string "@" ClientIP " logged in as " UserAgent:string "]"
+};
+let vCenter_Logout =() {vcenter_CL
+| where Message has ("UserLogoutSessionEvent")
+| parse Message with * "Event [" EventId:string "] [1-1] [" EventTime:datetime "] [" EventType:string "] [" EventSeverity "]" * "[User" DomainName:string "\\" Username:string "@" ClientIP " logged out (login time:" LoginTime:string ", number of API invocations: " APIInvocationCount:int ", user agent:" UserAgent:string ")]"
+};
+let vCenter_Role=() {vcenter_CL
+| where Message has_any("RoleAddedEvent","RoleRemovedEvent")
+| parse Message with eventdate:string " " eventtime:string " " Hostname:string "Event [" EventId:string "] [1-1] [" EventTime:datetime "] [" EventType:string "] [" EventSeverity:string "]" * "[New role " RoleName:string " " Operation:string  "]"
```

Comment:
Based on line 40, vpxd will also be present in this event. If we have to remove it from Hostname, it should be done here as well.

Reply:
Fixed